R/age_model_class.R

```diff
@@ -0,0 +1,222 @@
+    #add names to each value
+    names(params) = object@parameter_names
+
+    #raise errors if age category dimensions do not match initial state vectors
+    #also raise errors if initial state and parameter values are not doubles
+    for (p in list('S0', 'E0', 'I0', 'R0')){
+      if(length(params[[p]]) != object@n_age_categories){
+        stop(glue('Wrong number of age groups for {p}
+              compartments.'))}
+      if(!is.double(params[[p]])){
+        stop(glue('{p} storage format must be a vector.'))}
+    }
+    if(sum(S0, E0, I0, R0) != 1){
+      stop('All comparmtnets need to sum up to 1.')
+    }
+
+    #check format of parameters a and b
+    if(any(length(a) != 1 | length(b) != 1 | length(c) != 1)){
+      stop('The rates of change between compartments are 1-dimensional.')
+    }
+
+    #if all above tests are passed, assign the params namelist to the object
+    object@parameters <- params
+
+    return(object)
+  })
+
+setGeneric(name = 'simulate',
+           def = function(object, times = seq(0, 100, by = 1), is_plot=TRUE){
+             standardGeneric('simulate')
+           }
+)
+
+#' @describeIn Solves an age-structured simple SEIR model.
+#'
+#' Default time series is seq(0, 100, by = 1).
+#' This function relies on the packages deSolve and ggplot2.
+#' This function creates a plot of the variables over time and returns a
+#' vectors of the incidence numbers for each age group.
+#'
+#' @param object An object of the class age_model.
+#' @param times (vector) time sequence over which to solve the model.
+#'        Must be of the form seq(t_start,t_end,by=t_step).
+#' @param is_plot (logical) whether plots for the compartments for the
+#'        different age groups are drawn.
+#'
+#' @return data frame of time and outputs with incidence numbers for each
+#'         age group.
+setMethod(
+  'simulate', 'age_model',
+  function(object, times, is_plot) {
+    if(!is.double(times)){
+      stop('Evaluation times of the model storage format must be a vector.')
+    }
+    if(!is.logical(is_plot)){
+      stop('Trigger for plotting must be logical.')
+    }
+    age <- object@n_age_categories
+
+    #set initial state vector
+    state <- c(S = get_parameters(object)$S0,
+               E = get_parameters(object)$E0,
+               I = get_parameters(object)$I0,
+               R = get_parameters(object)$R0)
+
+    #set parameters vector
+    parameters <- c(a = get_parameters(object)$a,
+                    b = get_parameters(object)$b,
+                    c = get_parameters(object)$c)
+
+    #function for RHS of ode system
+    right_hand_side <- function(t, state, parameters) {
+      with(
+        as.list(c(state, parameters)),
+        {
+          S <- state[1:age]
+          E <- state[(age+1):(2*age)]
+          I <- state[(2*age+1):(3*age)]
+          R <- state[(3*age+1):(4*age)]
+          # rate of change
+          dS <- -a*S*I
+          dE <- a*S*I - b*E
+          dI <- b*E - c*I
+          dR <- c*I
```

Comment:
I think I'd be tempted to solve for cumulative incidence here via `dCdt = beta * S * I`. Could we also add deaths (from the infected compartment)?